draft-irtf-cfrg-hpke.md

```diff
@@ -273,50 +275,90 @@ values are two bytes long.
 
 ## DH-Based KEM
 
-Suppose we are given a Diffie-Hellman group that provides the
-following operations:
+Suppose we are given an instantiation KDF_kem of KDF with functions
+Hash_kem, Extract_kem, and Expand_kem, and a Diffie-Hellman group
+providing the following operations:
 
 - GenerateKeyPair(): Generate an ephemeral key pair `(sk, pk)`
   for the DH group in use
 - DH(sk, pk): Perform a non-interactive DH exchange using the
-  private key sk and public key pk to produce a shared secret
-  of length Npk
-- Marshal(pk): Produce a fixed-length byte string
+  private key sk and public key pk to produce a Diffie-Hellman
+  shared secret of length Ndh
+- Marshal(pk): Produce a byte string of length Npk
   encoding the public key `pk`
-- Unmarshal(enc): Parse a fixed-length byte string to recover a
+- Unmarshal(enc): Parse a byte string of length Npk to recover a
   public key
+- Nzz: The output size of the Hash_kem and Extract_kem functions
+  in bytes
 
-Then we can construct a KEM (which we'll call "DHKEM") in the
-following way:
+Then we can construct a KEM called `DHKEM(Curve, Hash_kem)` in the
+following way, where `Curve` denotes the Diffie-Hellman group and
+`Hash_kem` the hash function underlying KDF_kem:
 
 ~~~
+label_dh = "RFCXXXX DHKEM"
+
 def Encap(pkR):
   skE, pkE = GenerateKeyPair()
-  zz = DH(skE, pkR)
+  dh = DH(skE, pkR)
   enc = Marshal(pkE)
+
+  pkRm = Marshal(pkR)
+  context_kem = concat(enc, pkRm)
+
+  prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))
```

Comment:
These lines (prk = ..., zz = ...) are the same across all variants. Can we pull them out into a common function that takes `dh`, `context_kem` as input?

```
def ExtractAndExpand(dh, prk):
   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))
   zz  = Expand_kem(prk, concat("prk", context_kem), Nzz)
```

Then call like so:

```
def Encap(pkR):
   ...
   zz = ExtractAndExpand(dh, context_kem)
   return zz, enc
```